### ISSUE TITLE
<type_traits>: std::is_function causes ICE when used with function pointers from lambdas in VS 17.3.0 Preview 2+

### ISSUE THREAD

**Issue by marzer** [microsoft/STL#2986]
**Describe the bug**
See title. 

*Disclaimer:* I'm fully aware this is a compiler bug, my apologies! I reported it on devcomm the first day it appeared after installing 17.3.0 Preview 2 ([bug report here](https://developercommunity.visualstudio.com/t/Conversion-of-lambda-to-function-pointer/10071872)). It's 'fixed' but hasn't been released, and with each additional preview release I'm growing more worried it will leak into the final release of that version, rendering VS un-updateable for our team. I'm reporting it here in the hope that one of the Microsoft employees who work on this repository can escalate this internally, because the response there so far has been \*crickets\* :(

**Command-line test case**
```
C:\Temp>type repro1.cpp && cl /std:c++latest .\repro1.cpp
#include <type_traits>

int main()
{
	auto lambda = []() {};
	static_assert(std::is_function_v<std::remove_pointer_t<decltype(+lambda)>>);

	return 0;
}
Microsoft (R) C/C++ Optimizing Compiler Version 19.33.31629 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro1.cpp
.\repro1.cpp(6): fatal error C1001: Internal compiler error.
(compiler file 'msc1.cpp', line 1576)
 To work around this problem, try simplifying or changing the program near the locations listed above.
If possible please provide a repro here: https://developercommunity.visualstudio.com
Please choose the Technical Support command on the Visual C++
 Help menu, or open the Technical Support help file for more information
INTERNAL COMPILER ERROR in 'C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\bin\HostX86\x86\cl.exe'
    Please choose the Technical Support command on the Visual C++
    Help menu, or open the Technical Support help file for more information
```

If you hoist the function type out into a separate typedef you get a rather telling message about the calling convention being incorrect (which is I suspect the actual underlying cause):
```
C:\Temp>type repro2.cpp && cl /std:c++latest .\repro2.cpp
#include <type_traits>

int main()
{
	auto lambda = []() {};
	using func_type = std::remove_pointer_t<decltype(+lambda)>;
	static_assert(std::is_function_v<func_type>);

	return 0;
}
Microsoft (R) C/C++ Optimizing Compiler Version 19.33.31629 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro2.cpp
.\repro2.cpp(7): error C3865: '__thiscall': can only be used on native member functions
.\repro2.cpp(7): fatal error C1907: unable to recover from previous error(s); stopping compilation
INTERNAL COMPILER ERROR in 'C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31629\bin\HostX86\x86\cl.exe'
    Please choose the Technical Support command on the Visual C++
    Help menu, or open the Technical Support help file for more information
```

**Expected behavior**
The program compiles without error.

**STL version**
VS 17.3.0 Preview 2.0+ (still present in latest preview)

**Additional context**
Compiler Explorer: https://godbolt.org/z/37z345K86

**Comment by StephanTLavavej** [comment]
Thanks for bringing this issue to our attention. With the DevCom bug link you provided, I was able to track down the internal PR that fixed the bug. It will ship in VS 2022 17.4 Preview 1 (we can't promise an exact release date for this, but it'll be very soon). Unfortunately, even though the compiler team fixed this only 5 days after your bug was filed (June 15 reported, June 20 fixed), this fix didn't make it into the 17.3 release branch due to the immense latency in our release pipeline, so the 17.3 production release will contain the regression :crying_cat_face:

**Comment by marzer** [comment]
Thanks @StephanTLavavej! Sorry to spam this repo. Good to know the lag time won't be so bad. I can be noisy about not updating for one or two releases.
